core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -70,21 +76,25 @@ class LogSegment(val log: FileMessageSet,
    *
    * It is assumed this method is being called from within a lock.
    *
-   * @param offset The first offset in the message set.
+   * @param firstOffset The first offset in the message set.
+   * @param timestamp The largest timestamp in the message set.
    * @param messages The messages to append.
    */
   @nonthreadsafe
-  def append(offset: Long, messages: ByteBufferMessageSet) {
+  def append(firstOffset: Long, timestamp: Long, messages: ByteBufferMessageSet) {
     if (messages.sizeInBytes > 0) {
-      trace("Inserting %d bytes at offset %d at position %d".format(messages.sizeInBytes, offset, log.sizeInBytes()))
+      trace("Inserting %d bytes at offset %d at position %d".format(messages.sizeInBytes, firstOffset, log.sizeInBytes()))
+      val physicalPosition = log.sizeInBytes()
+      // append the messages
+      log.append(messages)
       // append an entry to the index (if needed)
       if(bytesSinceLastIndexEntry > indexIntervalBytes) {
-        index.append(offset, log.sizeInBytes())
-        this.bytesSinceLastIndexEntry = 0
+        index.append(firstOffset, physicalPosition)
+        timeIndex.maybeAppend(maxTimestampSoFar, firstOffset)
+        bytesSinceLastIndexEntry = 0
       }
-      // append the messages
-      log.append(messages)
-      this.bytesSinceLastIndexEntry += messages.sizeInBytes
+      maxTimestampSoFar = math.max(maxTimestampSoFar, timestamp)
```

Comment:
If we compute the maxTimestampSoFar before appending to the timeindex, we will require the lastOffset of the messages. If we do it in this way we only need the firstOffset.